upgrade/upgrade-scripts/rootfs-cleanup

```diff
@@ -0,0 +1,154 @@
+
+    def __lt__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "lt", version(other.rootfs))
+
+    def __gt__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "gt", version(other.rootfs))
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, rootfscmp):
+            return NotImplemented
+        return dpkgcmp(version(self.rootfs), "eq", version(other.rootfs))
+
+    def __le__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "le", version(other.rootfs))
+
+    def __ge__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "ge", version(other.rootfs))
+
+    def __ne__(self, other: object) -> bool:
+        if not isinstance(other, rootfscmp):
+            return NotImplemented
+        return dpkgcmp(version(self.rootfs), "ne", version(other.rootfs))
+
+
+def main() -> None:
+    snapshots = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "2", "-Ho", "name", "-t",
+        "snapshot"
+    ]).decode("UTF-8").splitlines()
+
+    for snap in snapshots:
+        name = snap.split("@")[1]
+
+        #
+        # We filter out any container-delphix snapshots, as those removed by
+        # the calls to "rootfs-container delete", on the relevant filesystems.
+        #
+        if re.search("^container-delphix.\w{7}$", name):
+            snapshots.remove(snap)
+            continue
+
+        #
+        # Out of an abundance of caution, verify the snapshot name matches
+        # our expectations.
+        #
+        if not re.search("^execute-upgrade.\w{7}$", name):
+            raise SystemExit("unexpected snapshot name: '{}'".format(name))
+
+    filesystems = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "1", "-Ho", "name", "-t",
+        "filesystem"
+    ]).decode("UTF-8").splitlines()
+
+    #
+    # We remove the first entry in the list, as that'll refer to the "rpool/ROOT"
+    # dataset, which is not versioned and also not eligible for deletion.
+    #
+    filesystems = filesystems[1:]
+
+    current = subprocess.check_output(["get-appliance-version"
+                                      ]).decode("UTF-8").strip()
+
+    #
+    # Our policy is to keep two delphix versions at any given time; the
+    # currently running version, and the version used prior to the
+    # current one.
+    #
+    # By sorting the list of all filesystems and snapshots on the sytem,
+    # we can easily determine the highest two versions on the system,
+    # and skip those versions (to maintain the policy described above);
+    # this is based on the assumption that the versions of an appliance
+    # will always increase (or remain constant, in the case of rollback)
+    # after an upgrade (i.e. we never upgrade to a lower version).
+    #
+    for rootfs in sorted(filesystems + snapshots, key=rootfscmp)[:-2]:
```

Comment:
wondering if it is not sufficient to sort the above zfs list commands by the "creation" field with "-s creation". Are there scenarios where the two orders (one by creation time and the other by associated delphix version) can be different?

Reply:
I thought about that, after I had implemented it this way. I think that would probably work too. You think I should try that way? I'm not sure if sorting by creation date would result it a cleaner implementation, but it might.. I can try it and see how I like it.

Reply:
I think sorting on creation date would work.. do you prefer that approach? vs. sorting on delphix version?

If we do that, should we still maintain the logic to skip versions greater than the current? i.e. this logic:
```
113         #                                                                       
114         # In the event of a rollback, we want to be careful to not              
115         # delete the currently running version, as well as any versions         
116         # that may have failed to upgrade resulting in the rollback             
117         # event to begin with. Thus, to account for the exceptional case        
118         # of rollback, we simply skip any version that's equal to or            
119         # greater than what's currently running.                                
120         #                                                                       
121         if dpkgcmp(version(rootfs), "ge", current):                             
122             break   
```
This assumes versions always increase, and I feel like the point of using the creation date is to remove that assumption.. so we should also remove the assumption from this condition (i.e. remove it entirely), but then we need to ensure the cleanup (based on creation date) works properly for rollback..

Reply:
@sanbhara Looking at this a little closer, I don't think the creation date would work for the filesystems.. Since, for those, the creation date will be the date it was originally created, which may have been at a much older version/time than what it's currently at..

So, if we wanted to use the creation date, I think we could only do that for snapshots, and would have to revert back to versions for the filesystems.. Since we can't use it for both, I'm not sure it's worth the effort (and/or inconsistency and complexity) to use it for one..

What do you think?

Reply:
Ok I see. Yes if we have to use the version property for one set of objects, we might as well do it for both sets of objects.

Reply:
OK, I'll go ahead and mark this comment as resolved.